python/tvm/topi/cuda/unique.py

```diff
@@ -0,0 +1,394 @@
+        index_converter_ptr = ib.buffer_ptr(index_converter)
+
+    if isinstance(counts, tir.Buffer):
+        counts_ptr = ib.buffer_ptr(counts)
+        arange_ptr = ib.allocate(counts_ptr.dtype, counts.shape, name="arange_buf", scope="global")
+
+    batch_size = data.shape[0]
+    max_threads = tir.min(batch_size, tvm.target.Target.current(allow_none=False).max_num_threads)
+
+    # calculate unique elements and inverse indices
+    with ib.new_scope():
+        nthread_tx = max_threads
+        nthread_bx = ceil_div(batch_size, max_threads)
+        tx = te.thread_axis("threadIdx.x")
+        bx = te.thread_axis("blockIdx.x")
+        ib.scope_attr(tx, "thread_extent", nthread_tx)
+        ib.scope_attr(bx, "thread_extent", nthread_bx)
+        tid = bx * max_threads + tx
+        with ib.if_scope(tid < batch_size):
+            data_idx = argsorted_indices_ptr[tid]
+            unique_idx = (
+                inc_scan_ptr[tid]
+                if not index_converter_ptr
+                else index_converter_ptr[inc_scan_ptr[tid]]
+            )
+            indices_ptr[data_idx] = unique_idx
+            with ib.if_scope(tid == 0):
+                unique_elements_ptr[unique_idx] = data_ptr[data_idx]
+            with ib.else_scope():
+                with ib.if_scope(inc_scan_ptr[tid] != inc_scan_ptr[tid - 1]):
+                    unique_elements_ptr[unique_idx] = data_ptr[data_idx]
+
+    # if need to return counts
+    if isinstance(counts, tir.Buffer):
+        num_unique = inc_scan_ptr[inc_scan.shape[0] - 1] + 1
+        num_elements = data.shape[0]
+        with ib.new_scope():
+            nthread_tx = max_threads
+            nthread_bx = ceil_div(batch_size, max_threads)
+            tx = te.thread_axis("threadIdx.x")
+            bx = te.thread_axis("blockIdx.x")
+            ib.scope_attr(tx, "thread_extent", nthread_tx)
+            ib.scope_attr(bx, "thread_extent", nthread_bx)
+            tid = bx * max_threads + tx
+            with ib.if_scope(tid < batch_size):
+                with ib.if_scope(tid == 0):
+                    arange_ptr[num_unique - 1] = num_elements
+                with ib.else_scope():
+                    with ib.if_scope(inc_scan_ptr[tid] != inc_scan_ptr[tid - 1]):
+                        arange_ptr[inc_scan_ptr[tid] - 1] = tid
+        with ib.new_scope():
+            nthread_tx = max_threads
+            nthread_bx = ceil_div(batch_size, max_threads)
+            tx = te.thread_axis("threadIdx.x")
+            bx = te.thread_axis("blockIdx.x")
+            ib.scope_attr(tx, "thread_extent", nthread_tx)
+            ib.scope_attr(bx, "thread_extent", nthread_bx)
+            tid = bx * max_threads + tx
+            with ib.if_scope(tid < num_unique):
+                unique_idx = tid if not index_converter_ptr else index_converter_ptr[tid]
+                with ib.if_scope(tid == 0):
+                    counts_ptr[unique_idx] = arange_ptr[tid]
+                with ib.else_scope():
```

Comment:
Probably we don't need to allocate `arange` array explicitly. Just assign `tid` should work?

Reply:
Sorry the name may be confusing. Actually the `arange_ptr` here records the cumulated counts of the unique elements and it goes like:
Step1: inc_scan = [0, 0, 1, 1, 2, 2, 2, 2, 2, 3]
                    tids = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
Step2: arange_ptr = [2, 4, 9, 10]
Step3: counts_ptr = [2, 2, 5, 1]

Maybe we can calculates `counts` first and use `indices_ptr` to replace `arange_ptr`?

Reply:
I didn't get why you need to fill in `arange_ptr` with cumulated counts or need a temp buffer. If you move the condition `ib.if_scope(inc_scan_ptr[tid] != inc_scan_ptr[tid - 1])` at L200 here, can't you inline the difference computation without materializing `arange_ptr`? I guess you will be writing at index `inc_scan_ptr[tid] - 1` instead of `unique_idx`.

Reply:
I'm imaging that we can have something like

`counts_ptr[inc_scan_ptr[tid] - 1] = tid - indices_ptr[inc_scan_ptr[tid - 1]]`

since the contents of `arange` seems redundant given other information you already have here, I have a strong feeling that we don't need to materialize `arange`.

Reply:
We can also consider filling in `counts` array just like you fill in `arange`, and do in-place `_calc_adjacent_diff` on it directly. 

Anyway, as long as we remove `arange` array, I'm happy.

Reply:
@ymwangg I made a lot of noise above, but after some thought I realized that probably we cannot remove `arange`: We cannot tell which previous index in arange would correspond to the previous unique element, without materializing `arange` array and doing global sync. In-place adjacent diff obvious doesn't work without some synchronization. 

So maybe we rename `arange` to something like `unique_seq_indices` and keep everything else as is, that would be good?

Reply:
Right, the main problem is for a `tid` with condition `inc_scan[tid] != inc_scan [tid]` we also need to know the previous `tid` with the same condition. It's possible to use `count` to store the cumulative sum but getting the adjacent diff in-place requires global sync.

I think we can use `indices_ptr` as a tmp buf to avoid allocating a new array.

Reply:
I'd hope that TVM could automatically do such buffer reuse optimization, but not sure if it would work in this case (can check by dumping the cuda source). I'm also ok with having `arange` explicitly for clarity.

Reply:
I checked the cuda kernels and if they are executed in the order they appear in IR (I think they do), then I don't see any potential issues.